### PR TITLE
Enhance State Transfer Status for concord-ctl

### PR DESCRIPTION
**Pull request by saiakhil2012**
Enhancing the State Transfer Status for `concord-ctl` by,
- Making the output JSON based
- Adding more details related to General State Transfer Metrics
- Adding more details when the replica is in Collecting State

Tested the code changes, by manually inducing a replica to go into state-transfer scenario

### CODE REVIEW

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 1]
```diff
@@ -35,7 +36,10 @@
 // TODO(GG): for debugging - remove
 // #define DEBUG_SEND_CHECKPOINTS_IN_REVERSE_ORDER (1)
 
+#define getName(var) #var
```
Comment (robem) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:39]:
```suggestion
#define VARSTR(var) #var
```
or
```suggestion
#define NAME2STR(var) #var
```

**Review by robem — APPROVED** [review]
Lgtm. Thanks for the enhancement!

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 1 above)
Comment (andrewjstone) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:39]: Looking at the rest of the code, I see that we only use this when we call `toPair(getName(var), var)`. That gets quite repetitive. I suggest turning the whole thing into a macro such that you just write `STRPAIR(var)` or something like that.

**Review by andrewjstone — APPROVED** [review]
This is great! Thanks @saiakhil2012. I only have one minor suggestion.

**Review by glevkovich — APPROVED** [review]
Looks good @saiakhil2012.
Just one comment: we should consider adding also the FetchBlockMsg <-> ItemDataMsg latencies. This is recorded by the BCStateTran::histograms_.
It is already possible to get it with concord-ctl from the command line, by calling:
> concord-ctl perf snapshot state_transfer
and then 
>concord-ctl perf get fetch_blocks_msg_latency
But maybe it can be nice to have it printed as part of the full status update?

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 1 above)
Comment (saiakhil2012) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp]: Yeah, I agree, the use of `toPair(getName(var), var)` is quite repetitive here. Sure, as suggested, I moved to using macro. Thanks

**Comment by saiakhil2012** [comment]
> Looks good @saiakhil2012.
> Just one comment: we should consider adding also the FetchBlockMsg <-> ItemDataMsg latencies. This is recorded by the BCStateTran::histograms_.
> It is already possible to get it with concord-ctl from the command line, by calling:
> 
> ```
> concord-ctl perf snapshot state_transfer
> concord-ctl perf get fetch_blocks_msg_latency
> ```
> 
> But maybe it can be nice to have it printed as part of the full status update?

@glevkovich , I understand it would be good to have all together, but as the same content is already available via `perf` commands of `concord-ctl`, then instead of we repeating content across multiple commands, we can combine output of multiple commands like, `{command1 & command2;}`. In this way, we can combine any histogram with any status on need basis while performing diagnostics. Also, apart from the above thought, the latencies would be in the form of histogram which is what we are providing in `perf` commands, whereas in `status` we are focusing on status which is more of key-value pair format? Any thoughts?

**Comment by teoparvanov** [comment]
> 
> 
> > Looks good @saiakhil2012.
> > Just one comment: we should consider adding also the FetchBlockMsg <-> ItemDataMsg latencies. This is recorded by the BCStateTran::histograms_.
> > It is already possible to get it with concord-ctl from the command line, by calling:
> > ```
> > concord-ctl perf snapshot state_transfer
> > concord-ctl perf get fetch_blocks_msg_latency
> > ```
> > 
> > 
> > But maybe it can be nice to have it printed as part of the full status update?
> 
> @glevkovich , I understand it would be good to have all together, but as the same content is already available via `perf` commands of `concord-ctl`, then instead of we repeating content across multiple commands, we can combine output of multiple commands like, `{command1 & command2;}`. In this way, we can combine any histogram with any status on need basis while performing diagnostics. Also, apart from the above thought, the latencies would be in the form of histogram which is what we are providing in `perf` commands, whereas in `status` we are focusing on status which is more of key-value pair format? Any thoughts?

I agree with @saiakhil2012 , let's not duplicate the information here.

**Review by teoparvanov — APPROVED** [review]
LGTM, thank you @saiakhil2012 !

**Pull request merged**